### PR TITLE
Generate new buffer/get buffer create

### PR DESCRIPTION
**Pull request by josephmturner**
For testing, would it make sense to repurpose the following from `/src/buffer-tests.el`?

```
(ert-deftest buffer-tests-inhibit-buffer-hooks ()
  "Test `get-buffer-create' argument INHIBIT-BUFFER-HOOKS."
  (let* (run-bluh (bluh (lambda () (setq run-bluh t))))
    (unwind-protect
        (let* ( run-kbh  (kbh  (lambda () (setq run-kbh  t)))
                run-kbqf (kbqf (lambda () (setq run-kbqf t))) )

          ;; Inhibited.
          (add-hook 'buffer-list-update-hook bluh)
          (with-current-buffer (generate-new-buffer " foo" t)
            (add-hook 'kill-buffer-hook kbh nil t)
            (add-hook 'kill-buffer-query-functions kbqf nil t)
            (kill-buffer))
          (with-temp-buffer (ignore))
          (with-output-to-string (ignore))
          (should-not run-bluh)
          (should-not run-kbh)
          (should-not run-kbqf)

          ;; Not inhibited.
          (with-current-buffer (generate-new-buffer " foo")
            (should run-bluh)
            (add-hook 'kill-buffer-hook kbh nil t)
            (add-hook 'kill-buffer-query-functions kbqf nil t)
            (kill-buffer))
          (should run-kbh)
          (should run-kbqf))
      (remove-hook 'buffer-list-update-hook bluh))))

(ert-deftest buffer-tests-inhibit-buffer-hooks-indirect ()
  "Test `make-indirect-buffer' argument INHIBIT-BUFFER-HOOKS."
  (let* ( base run-bluh run-kbh run-kbqf
          (bluh (lambda () (setq run-bluh t)))
          (kbh  (lambda () (setq run-kbh  t)))
          (kbqf (lambda () (setq run-kbqf t))))
    (dolist (inhibit-base '(nil t))
      (unwind-protect
          (let (indirect)
            (setq base (generate-new-buffer " base" inhibit-base))
            (dolist (inhibit-indirect '(nil t))
              (dotimes (_ 11)
                (unwind-protect
                    (let ((name (generate-new-buffer-name " indirect")))
                      (setq run-bluh nil run-kbh nil run-kbqf nil)
                      (add-hook 'buffer-list-update-hook bluh)
                      (with-current-buffer
                          (setq indirect (make-indirect-buffer
                                          base name nil inhibit-indirect))
                        (add-hook 'kill-buffer-hook kbh nil t)
                        (add-hook 'kill-buffer-query-functions kbqf nil t)
                        (kill-buffer))
                      (should (xor inhibit-indirect run-bluh))
                      (should (xor inhibit-indirect run-kbh))
                      (should (xor inhibit-indirect run-kbqf)))
                  (remove-hook 'buffer-list-update-hook bluh)
                  (when (buffer-live-p indirect)
                    (kill-buffer indirect))))))
        (when (buffer-live-p base)
          (kill-buffer base))))))
```

### CODE REVIEW

File: compat-28.el [hunk 1]
```diff
@@ -392,6 +392,39 @@ not a list, return a one-element list containing OBJECT."
...
+presented to users or passed on to other applications.
+
+If BUFFER-OR-NAME is a buffer instead of a string, return it as given,
+even if it is dead.  The return value is never nil."
+  (if inhibit-buffer-hooks
+      (let (kill-buffer-hook
+            kill-buffer-query-functions
+            buffer-list-update-hook)
+        (get-buffer-create name))
+    (get-buffer-create name inhibit-buffer-hooks)))
```
Comment (minad) [inline compat-28.el:415]: The argument inhibit-buffer-hooks should be removed here.

File: compat-28.el [hunk 2]
```diff
@@ -392,6 +392,39 @@ not a list, return a one-element list containing OBJECT."
...
+
+If BUFFER-OR-NAME starts with a space, the new buffer does not keep undo
+information.  If optional argument INHIBIT-BUFFER-HOOKS is non-nil, the
+new buffer does not run the hooks `kill-buffer-hook',
+`kill-buffer-query-functions', and `buffer-list-update-hook'.  This
+avoids slowing down internal or temporary buffers that are never
+presented to users or passed on to other applications.
+
+If BUFFER-OR-NAME is a buffer instead of a string, return it as given,
+even if it is dead.  The return value is never nil."
```
Comment (minad) [inline compat-28.el:409]: The function must specify the keyword `:extended t` right after the docstring.

File: compat-28.el [hunk 3]
```diff
@@ -392,6 +392,39 @@ not a list, return a one-element list containing OBJECT."
...
+            kill-buffer-query-functions
+            buffer-list-update-hook)
+        (get-buffer-create name))
+    (get-buffer-create name inhibit-buffer-hooks)))
+
+(compat-defun generate-new-buffer (name &optional inhibit-buffer-hooks)
+  "Create and return a buffer with a name based on NAME.
+Choose the buffer's name using `generate-new-buffer-name'.
+See `get-buffer-create' for the meaning of INHIBIT-BUFFER-HOOKS."
+  (if inhibit-buffer-hooks
```
Comment (minad) [inline compat-28.el:421]: The function must specify the keyword :extended t right after the docstring.

**Comment by minad** [comment]
> For testing, would it make sense to repurpose the following from /src/buffer-tests.el?

Yes, these tests seem fine. Please add them. Also update compat.texi, the section "Extended functions" of Emacs 28.

File: compat-28.el [hunk 4]
```diff
@@ -392,6 +392,39 @@ not a list, return a one-element list containing OBJECT."
       object
     (list object)))
 
+(compat-defun get-buffer-create (name &optional inhibit-buffer-hooks)
+  "Return the buffer specified by BUFFER-OR-NAME, creating a new one if needed.
```
Comment (minad) [inline compat-28.el:396]: The doc string can be shorter here. It is sufficient to write only "This extended function implements the INHIBIT-BUFFER-HOOKS argument.", such that it is obvious that the function only adds new functionality over the existing `get-buffer-create`. See the other doc strings of `:extended t` functions for reference.

File: compat-28.el (hunk 1 above)
Comment (josephmturner) [reply compat-28.el:415]: Woops!  Thank you.

**Comment by josephmturner** [comment]
> Yes, these tests seem fine. Please add them. Also update compat.texi, the section "Extended functions" of Emacs 28.

I've copied the `@defun` sections for the two functions from `buffers.texi` and changed `@defun` -> `@defun compat-call@`.

Neither of those two tests directly call `get-buffer-create`, but I've linked the `compat-defun` for `get-buffer-create` to one of them to appease the CI.  Is that alright?

The two new tests pass on my local Emacs `29.3`, but the CI is failing.  I'd like to see if the CI passes on `29`, but those tests were skipped in the CI.

**Comment by phikal** [comment]
Joseph Turner ***@***.***> writes:

>> Yes, these tests seem fine. Please add them. Also update compat.texi, the section "Extended functions" of Emacs 28.
>
> I've copied the ***@***.***` sections for the two functions from `buffers.texi` and changed ***@***.***` -> ***@***.*** compat-call@`.

Can you attribute that with a comment before the @defun block?

> Neither of those two tests directly call `get-buffer-create`, but I've
> linked the `compat-defun` for `get-buffer-create` to one of them to
> appease the CI.  Is that alright?
>
> The two new tests pass on my local Emacs `29.3`, but the CI is failing.  I'd like to see if the CI passes on `29`, but those tests were skipped in the CI.

The tests for 29.1 were interrupted but still managed to print

    Ran 193 tests, 193 results as expected, 0 unexpected (2024-05-27 04:48:53+0000, 0.387917 sec)

-- 
	Philip Kaludercic on peregrine

**Comment by josephmturner** [comment]
> Can you attribute that with a comment before the @Defun block?

Done.

>

> The tests for 29.1 were interrupted but still managed to print Ran 193 tests, 193 results as expected, 0 unexpected (2024-05-27 04:48:53+0000, 0.387917 sec)
> […](#)

Thanks, I see that now.  Does this mean that the `let`-bind approach doesn't work?

File: compat-28.el [hunk 5]
```diff
@@ -392,6 +392,26 @@ not a list, return a one-element list containing OBJECT."
       object
     (list object)))
 
+(compat-defun get-buffer-create (name &optional inhibit-buffer-hooks) ;; <compat-tests:buffer-tests-inhibit-buffer-hooks>
```
Comment (minad) [inline compat-28.el:395]: Our convention is to have separate tests for `get-buffer-create` and `generate-new-buffer`. They should be named `compat-get-buffer-create` and `compat-generate-new-buffer`.

File: compat-tests.el [hunk 6]
```diff
@@ -1004,6 +1004,64 @@
       (should-equal compat-tests--global 1)
       (should-equal compat-tests--local 2)
       (should-not (boundp 'compat-tests--unexist)))))
+ 
+(ert-deftest compat-buffer-tests-inhibit-buffer-hooks ()
```
Comment (minad) [inline compat-tests.el:1008]: I suggest to merge the two complex tests into one test function `compat-generate-new-buffer`. Then add a simpler separate test `compat-get-buffer-create`. I insist on having separate tests for each Compat-provided function.

**Comment by minad** [comment]
@josephmturner 

> Thanks, I see that now. Does this mean that the let-bind approach doesn't work?

I haven't understood why the tests don't succeed. My suggestion is to get it to work locally with old Emacs 27 in addition to either Emacs 28, 29 or 30. For 27 your compatibility code will be used, while for the newer Emacs versions, the builtin functions are used. If this works, we will have a good chance to get it to work. Maybe there are other subtleties on even older version, but these can usually get resolved.

**Comment by josephmturner** [comment]
> I haven't understood why the tests don't succeed. My suggestion is to get it to work locally with old Emacs 27 in addition to either Emacs 28, 29 or 30. For 27 your compatibility code will be used, while for the newer Emacs versions, the builtin functions are used. If this works, we will have a good chance to succeed. Maybe there are other subtleties on even older version, but these can usually get resolved.

Working on it now, please see [this issue with guix-emacs-historical](https://lists.sr.ht/~pkal/public-inbox/%3C87ttiiq0u8.fsf@breatheoutbreathe.in%3E).

File: compat-tests.el (hunk 6 above)
Comment (josephmturner) [reply compat-tests.el:1008]: What did you have in mind for a test for `compat-get-buffer-create`? Would the test for `compat-get-buffer-create` be modeled off of the test for `compat-generate-new-buffer`?

File: compat-28.el [hunk 7]
```diff
@@ -392,6 +392,26 @@ not a list, return a one-element list containing OBJECT."
...
+ 
+(compat-defun generate-new-buffer (name &optional inhibit-buffer-hooks) ;; <compat-tests:buffer-tests-inhibit-buffer-hooks-indirect>
+  "Handle additional optional argument INHIBIT-BUFFER-HOOKS."
+  :extended t
+  (if inhibit-buffer-hooks
+      (let (kill-buffer-hook
+            kill-buffer-query-functions
+            buffer-list-update-hook)
+        (get-buffer-create (generate-new-buffer-name name)))
+    (get-buffer-create (generate-new-buffer-name name))))
```
[the remark this answers is not in the document]
Comment (josephmturner) [reply compat-28.el]: Done, thanks!

File: compat-tests.el (hunk 6 above)
Comment (minad) [reply compat-tests.el:1008]: No, a simple test would be fine, given that we already test the function indirectly, i.e., just calling the function once with and once without the argument, checking that it returns a buffer with a new name. I want to make sure that the function is defined via Compat. That's all.

**Comment by josephmturner** [comment]
@minad [This SourceHut thread](https://lists.sr.ht/~pkal/public-inbox/%3C87ttiiq0u8.fsf@breatheoutbreathe.in%3E) concludes that `guix-emacs-historical` doesn't currently work for running old versions of Emacs.

I'm on Debian 12 (bookworm).  How would you recommend running Emacs 27?

**Comment by alphapapa** [comment]
> I'm on Debian 12 (bookworm). How would you recommend running Emacs 27?

Here are a few ideas:

1. Use `makem.sh` and its GitHub CI integration to run tests on Emacs 27 on GitHub CI.  For non-interactive, non-GUI-relevant tests, this should work, even if it's not necessarily as convenient for development (but once developed, the tests can continue running on CI).
2. I'd be surprised if there's not a Docker image that provides Emacs 27.
3. Use a VM with the last version of Debian that provides Emacs 27.
4. Or you could build it yourself, as mentioned.  Note that you don't have to install that version of Emacs after building it; you can run it from the build directory.

**Comment by josephmturner** [comment]
I'm noticing that on Emacs 29, if I attempt to run the `compat-buffer-tests-inhibit-buffer-hooks` test defined in compat-tests.el on this branch, it fails if instrumented for edebug.

**Comment by josephmturner** [comment]
Apparently, I was wrong about `buffer-list-update-hook` being introduced in Emacs 28.  In the VM I'm using to test Emacs 27, `buffer-list-update-hook` is defined.

**Comment by josephmturner** [comment]
I am not sure that it is possible to add this feature to `compat`. `let`-binding the hooks to nil as in the proposed patch doesn't work because the `let` bindings only apply during the call to `get-buffer-create`.  However, the expected behavior is that the three global hooks will not be run when the buffer is created or killed.  For this, we could perhaps modify each existing function in each hook to do nothing if the buffer being created, renamed, or killed, etc. is the current buffer.

Already this is too complex, and we haven't accounted for new functions added to any of these hooks after the call to `get-buffer-create`.

**Comment by minad** [comment]
Yes, I also have the feeling that we cannot make this work without intrusive changes, which are out of scope of Compat.

**Comment by josephmturner** [comment]
Thanks for exploring this idea!

**Pull request closed without merging**